src/verifiers/documentStatus/tokenRegistryStatus/openAttestationEthereumTokenRegistryStatus.ts

```diff
@@ -1,22 +1,91 @@
 import { getData, utils, v2, v3, WrappedDocument } from "@govtechsg/open-attestation";
 import { TradeTrustErc721Factory } from "@govtechsg/token-registry";
-import { constants } from "ethers";
+import { constants, errors } from "ethers";
 import { VerificationFragmentType, Verifier } from "../../../types/core";
 import { OpenAttestationEthereumTokenRegistryStatusCode } from "../../../types/error";
-import { contractNotMinted, getErrorReason } from "./errors";
-import { getIssuersTokenRegistry, getProvider } from "../../../common/utils";
+import { getProvider } from "../../../common/utils";
+import { withCodedErrorHandler } from "../../../common/errorHandler";
+import { CodedError } from "../../../common/error";
 
 interface Status {
   minted: boolean;
   address: string;
   reason?: any;
 }
 
-const isWrappedV2Document = (document: any): document is WrappedDocument<v2.OpenAttestationDocument> => {
-  return document.data && document.data.issuers;
-};
 const name = "OpenAttestationEthereumTokenRegistryStatus";
 const type: VerificationFragmentType = "DOCUMENT_STATUS";
+
+export const getTokenRegistry = (
+  document: WrappedDocument<v2.OpenAttestationDocument> | WrappedDocument<v3.OpenAttestationDocument>
+): string => {
+  if (utils.isWrappedV2Document(document)) {
+    const { issuers } = getData(document);
+    if (issuers.length !== 1)
+      throw new CodedError(
+        "Only one issuer is allowed for tokens",
+        OpenAttestationEthereumTokenRegistryStatusCode.INVALID_ISSUERS,
+        OpenAttestationEthereumTokenRegistryStatusCode[OpenAttestationEthereumTokenRegistryStatusCode.INVALID_ISSUERS]
+      );
+    if (!issuers[0].tokenRegistry)
+      throw new CodedError(
+        "Token registry is undefined",
+        OpenAttestationEthereumTokenRegistryStatusCode.UNDEFINED_TOKEN_REGISTRY,
+        OpenAttestationEthereumTokenRegistryStatusCode[
+          OpenAttestationEthereumTokenRegistryStatusCode.UNDEFINED_TOKEN_REGISTRY
+        ]
+      );
+    return issuers[0].tokenRegistry;
+  } else {
+    const { proof } = getData(document);
+    if (proof.method !== "TOKEN_REGISTRY")
+      throw new CodedError(
+        "Cannot validate non-token registry documents",
+        OpenAttestationEthereumTokenRegistryStatusCode.INVALID_VALIDATION_METHOD,
+        OpenAttestationEthereumTokenRegistryStatusCode[
+          OpenAttestationEthereumTokenRegistryStatusCode.INVALID_VALIDATION_METHOD
+        ]
+      );
+    return proof.value;
+  }
+};
+
+export const isTokenMintedOnRegistry = async ({
+  tokenRegistry,
+  merkleRoot,
+  network,
+}: {
+  tokenRegistry: string;
+  merkleRoot: string;
+  network: string;
+}): Promise<{ minted: boolean; reason?: string }> => {
+  try {
+    const tokenRegistryContract = await TradeTrustErc721Factory.connect(tokenRegistry, getProvider({ network }));
```

Comment:
- why getting rid of code and codeString ? that makes it a breaking change ...
- why getting rid of SERVER_ERROR ? We use it on opencert to differentiate a business and a technical error

Reply:
I knew you will ask, which is why i got you to review. 
For the errors which this verifier *knows for sure* that the result is either true or false, it returns that. 
For errors which it is uncertain if the document is issued or not, it will throw. 
There is no meaning in returning more information than that, oa-verify is meant to provide that level of abstraction to dependant applications.
In the cases from line 69 to 82, we are certain the token was not minted (regardless of which error it is), there is **no error** in those cases. 

I was initially reluctant to return more information that true/false.

Reply:
As for `SERVER_ERROR`, I can make the change to throw that one with a code.